org-roam-extra.el

```diff
@@ -0,0 +1,268 @@
+                tp (s-replace (car l) (nth 1 l) tp))))
+      (setf (nth 4 template) tp))
+    template))
+
+(defun org-roam-extra--switch-perspective ()
+  "Helper function for `org-roam-extra-org-ref-note-fn'."
+  (when
+      (and (featurep 'projectile)
+           (require 'projectile nil t)
+           (featurep 'persp-mode)
+           (require 'persp-mode nil t))
+    (let ((notes-project (cdr org-roam-extra-persp-project))
+          (projects (projectile-relevant-open-projects))
+          openp)
+      (dolist (project projects openp)
+        (setq openp (or (f-equal? project notes-project) openp)))
+      (when openp
+        (let ((p-names (cdr (persp-names))))
+          (dolist (p-name p-names)
+            (when (s-equals? p-name (car org-roam-extra-persp-project))
+              (persp-switch p-name))))))))
+
+;;;###autoload
+(defun org-roam-extra-org-ref-notes-fn (citekey)
+  "Open an org-roam note associated with the CITEKEY or create a new one.
+Set `org-ref-notes-function' to this function if your
+bibliorgaphy notes are managed by org-roam and you want some extra
+integration between the two packages.
+
+This function allows to use org-roam as a backend for org-ref's
+handling of bibliography notes.
+
+Implementation details and features:
+
+1. This funciton first calls `org-roam-find-ref' trying to find the
+note file associated with the CITEKEY. The org-roam key can be
+set with \'#+ROAM_KEY:\' in-buffer keyword.
+
+2. If the org-roam reference has not been found, the function
+calls `org-roam-find-file' passing to it the title associated
+with the CITEKEY as retrieved by `bibtex-completion-get-entry'.
+The prompt presented by `org-roam-find-file' will thus be
+pre-populated with the record title.
+
+3. Optionally, when `org-roam-extra-preformat-templates' is
+non-nil, any prompt wildcards in `org-roam-capture-templates'
+associated with the bibtex record fields as specified in
+`org-roam-extra-preformat-templates' will be preformatted. Both
+`org-capture-templates' (%^{}) and
+`org-roam-capture-templates' (`s-format', ${}) prompt syntaxes
+are supported.
+
+See `org-roam-extra-preformat-keywords' for more details on how
+to properly specify prompts for replacement.
+
+Please pay attention when using this feature that by setting
+title for preformatting it will be impossible to change it in the
+`org-roam-find-file' interactive prompt since all the template
+expansions will have taken place by then. All the title wildcards
+will be replace with the bibtex field value.
+
+4. Optionally, if you are using projectile and persp-mode and
+have a dedicated workspace to work with your org-roam collection,
+you may want to set the perspecive name and project path in
+`org-roam-extra-persp-project' and `org-roam-extra-switch-persp' to
+t. In this case, the perspective will be switched to the org-roam
+notes project before calling any org-roam functions."
+  (when
+      (and (featurep 'org-ref)
+           (require 'org-ref))
```

Comment:
Same here.

Reply:
I had to move `(require 'org-ref)` into the function body because the project would not pass build check otherwise, see above commit d79b245.

This, to my opinion, doesn't make much sense because it is actually a plugin function for org-ref and it should of course be loaded by then. So `(require 'org-ref)` must be either a top-level for or omitted altogether.